### ISSUE TITLE
array_merge: splat operator discards type information

### ISSUE THREAD

**Issue by weirdan** [vimeo/psalm#571]
```php
<?php
/**
 * @return array<int,array<int,string>>
 */
function getData(): array
{
    return [
        ["a", "b"],
        ["c", "d"]
    ];
}

/**
 * @return array<int,string>
 */
function f1(): array
{
    $data = getData();
    return array_merge($data[0], $data[1]);
}

/**
 * @return array<int,string>
 */
function f2(): array
{
    $data = getData();
    return array_merge(...$data);
}

/**
 * @return array<int,string>
 */
function f3(): array
{
    $data = getData();
    return array_merge([], ...$data);
}
```
Here all `f*` functions are equivalent, but psalm fails on `f2` and `f3`, thinking `array_merge()` would return `array<mixed,mixed>`. `f1` checks fine (as should the other two).

**Comment by weirdan** [comment]
It might be more general problem with splat operator: psalm does not complain about the following code (second parameter is passed as int, but should be string). `variadic($q[0], $q[1]);` is properly complained about.
```php
<?php
function variadic(string ...$args): void {}

$q = ["a", 1];
variadic(...$q);
```

**Comment by muglug** [comment]
Thanks! Turned out to be a very simple off-by-one error.
